packages/request-tracker/lib/request-tracker.ts

```diff
@@ -31,16 +31,23 @@ export class RequestTracker {
   }
 
   start (context: RequestStartContext) {
-    const endCallbacks: RequestEndCallback[] = []
+    const results: Array<ReturnType<RequestStartCallback>> = []
     for (const startCallback of this.callbacks) {
-      const endCallback = startCallback(context)
-      if (endCallback) endCallbacks.push(endCallback)
+      const result = startCallback(context)
+      if (result) results.push(result)
     }
 
-    return (endContext: RequestEndContext) => {
-      for (const endCallback of endCallbacks) {
-        endCallback(endContext)
-      }
+    return {
+      onRequestEnd: (endContext: RequestEndContext) => {
+        for (const result of results) {
+          result?.onRequestEnd?.(endContext)
```

Comment:
slight nit: I think we've been trying to avoid using optional chaining because it transpiles to a lot more code, e.g.
`(_a = result === null || result === void 0 ? void 0 : result.onRequestEnd) === null || _a === void 0 ? void 0 : _a.call(result, endContext)`